src/nginx/t/grpc_interop_upstream_ssl.t

```diff
@@ -0,0 +1,138 @@
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+################################################################################
+#
+use strict;
+use warnings;
+
+################################################################################
+
+use src::nginx::t::ApiManager;   # Must be first (sets up import path to the Nginx test module)
+use src::nginx::t::HttpServer;
+use src::nginx::t::ServiceControl;
+use Test::Nginx;  # Imports Nginx's test module
+use Test::More;   # And the test framework
+use JSON::PP;
+
+################################################################################
+
+# Port assignment
+my $Http2NginxPort = ApiManager::pick_port();
+my $ServiceControlPort = ApiManager::pick_port();
+my $GrpcBackendPort = ApiManager::pick_port();
+my $HttpBackendPort = ApiManager::pick_port();
+
+my $t = Test::Nginx->new()->has(qw/http proxy/)->plan(5);
+
+$t->write_file(
+    'service.pb.txt',
+    ApiManager::get_grpc_interop_service_config . <<"EOF");
+control {
+  environment: "http://127.0.0.1:${ServiceControlPort}"
+}
+EOF
+
+$t->write_file('root.crt', ApiManager::read_test_file('testdata/grpc/cacert.pem'));
+my $root_crt_path = $t->testdir() . '/root.crt';
+$t->write_file('server_config.pb.txt', <<"EOF");
+grpc_backend_ssl_credentials {
+  use_ssl: true
+  root_certs_file: "${root_crt_path}"
+}
+EOF
+
+$t->write_file_expand('nginx.conf', <<"EOF");
+%%TEST_GLOBALS%%
+daemon off;
+events {
+  worker_connections 32;
+}
+http {
+  %%TEST_GLOBALS_HTTP%%
+  server {
+    listen 127.0.0.1:${Http2NginxPort} http2;
+    server_name localhost;
+    location / {
+      endpoints {
+        api service.pb.txt;
+        server_config server_config.pb.txt;
+        on;
+      }
+      grpc_pass 127.0.0.1:${GrpcBackendPort};
+    }
+  }
+}
+EOF
+
+$t->write_file('test.key', ApiManager::read_test_file('testdata/grpc/serverkey.pem'));
+$t->write_file('test.crt', ApiManager::read_test_file('testdata/grpc/servercert.pem'));
+my @ssl_args = (
+  '--use_tls',
+  '--tls_key_file', $t->testdir() . '/test.key',
+  '--tls_cert_file', $t->testdir() . '/test.crt',
+);
+
+$t->run_daemon(\&service_control, $t, $ServiceControlPort, 'servicecontrol.log');
+$t->run_daemon(\&ApiManager::grpc_interop_server, $t, "${GrpcBackendPort}", @ssl_args);
+is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+is($t->waitforsocket("127.0.0.1:${GrpcBackendPort}"), 1, 'GRPC test server socket ready.');
+$t->run();
+is($t->waitforsocket("127.0.0.1:${Http2NginxPort}"), 1, 'Nginx socket ready.');
+
+################################################################################
+my @test_cases = (
```

Comment:
should there be any negative tests?

Reply:
Done.  Try to add two negative test cases, only one of them work, another one discovered a bug in grpc c++ code, have to disable that test.